eng/bomgenerator/src/main/java/com/azure/tools/bomgenerator/DependencyAnalyzer.java

```diff
@@ -0,0 +1,207 @@
+package com.azure.tools.bomgenerator;
+
+import com.azure.tools.bomgenerator.models.BomDependency;
+import com.azure.tools.bomgenerator.models.BomDependencyComparator;
+import com.azure.tools.bomgenerator.models.BomDependencyNoVersion;
+import com.azure.tools.bomgenerator.models.BomDependencyNonVersionComparator;
+import org.jboss.shrinkwrap.resolver.api.maven.Maven;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenArtifactInfo;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenFormatStage;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenResolvedArtifact;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenResolverSystemBase;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenStrategyStage;
+import org.jboss.shrinkwrap.resolver.api.maven.PomEquippedResolveStage;
+import org.jboss.shrinkwrap.resolver.api.maven.PomlessResolveStage;
+import org.jboss.shrinkwrap.resolver.api.maven.ScopeType;
+import org.jboss.shrinkwrap.resolver.api.maven.coordinate.MavenDependency;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import static com.azure.tools.bomgenerator.Utils.RESOLVED_EXCLUSION_LIST;
+
+public class DependencyAnalyzer {
+    private TreeSet<BomDependency> inputDependencies = new TreeSet<>(new BomDependencyComparator());
+    private TreeSet<BomDependency> externalDependencies = new TreeSet<>(new BomDependencyComparator());
+    private TreeSet<BomDependency> bomEligibleDependencies = new TreeSet<>(new BomDependencyComparator());
+    private TreeSet<BomDependency> bomIneligibleDependencies = new TreeSet<>(new BomDependencyComparator());
+
+    private TreeMap<BomDependencyNoVersion, HashMap<String, Collection<BomDependency>>> nameToVersionToChildrenDependencyTree = new TreeMap<>(new BomDependencyNonVersionComparator());
+    private static Logger logger = LoggerFactory.getLogger(BomGenerator.class);
+
+    DependencyAnalyzer(TreeSet<BomDependency> inputDependencies, TreeSet<BomDependency> externalDependencies){
+        if(inputDependencies != null) {
+            this.inputDependencies.addAll(inputDependencies);
+        }
+        if(externalDependencies != null) {
+            this.externalDependencies.addAll(externalDependencies);
+        }
+    }
+
+    public TreeSet<BomDependency> analyze() {
+        resolveTree();
+        filterConflicts();
+
+        return this.bomEligibleDependencies;
+    }
+
+    public Boolean validate() {
+        resolveTree();
+        return filterConflicts();
+    }
```

Comment:
I'm a bit confused, but doesn't `analyze` and `validate` do the same thing except their return types are different?

Reply:
Not really but based on your comment and I updated the method names and decoupled things a little

In the new one I have the following
1. AnalyzeAndReduce - The first builds the tree and removes the conflict nodes from the n-ary tree.
2. AnalyzeAndValidate - This builds the tree and tells whether there are conflicts in it or not.

The things we have to do to answer both are similar. Hope that clarifies.